File: apps/mobile/src/agent/planner.ts
import Constants from "expo-constants";
import { resolveOpenAIApiKey } from "@/services/ai";
import type { EnrichedPayload, PlannerResponse } from "@/agent/types";
import type { GoalContextItem } from "@/types/goal";
import { EdgeCache } from "@/agent/cache";

export interface PlannerResult {
  response: PlannerResponse | null;
  raw: any;
}

const PLANNER_SYSTEM_PROMPT = `Planner chooses tools: journal.create|goal.create|schedule.create|reflect|settings.update|noop.
Provided state includes structured goals, schedule windows, and constraints.
Output JSON is schema-enforced and idempotent. Never hallucinate IDs.`;

const ACTION_SCHEMA = {
  type: "object",
  additionalProperties: false,
  required: ["action", "ask", "payload"],
  properties: {
    action: {
      type: "string",
      enum: [
        "journal.create",
        "goal.create",
        "schedule.create",
        "reflect",
        "settings.update",
        "noop",
      ],
    },
    ask: {
      anyOf: [{ type: "string" }, { type: "null" }],
    },
    payload: {
      type: "object",
    },
  },
} as const;

const TOOL_DEFINITION = {
  type: "function",
  function: {
    name: "plan_router",
    description:
      "Select downstream action and structured payload for Riflett subsystems.",
    parameters: ACTION_SCHEMA,
  },
} as const;

const getExpoExtra = () =>
  (Constants?.expoConfig?.extra ?? {}) as Record<string, any>;

const resolveModel = (): string => {
  const extra = getExpoExtra();
  const configured = extra?.plannerModel as string | undefined;
  return configured || "gpt-4o-mini";
};

const summarizeGoalContext = (goals?: GoalContextItem[] | null): string => {
  if (!goals || goals.length === 0) {
    return "None";
  }

  const compact = goals.map((goal) => ({
    id: goal.id,
    title: goal.title,
    status: goal.status,
    priority: Number(goal.priority_score ?? 0).toFixed(2),
    current_step: goal.current_step,
    next_micro_steps: goal.micro_steps
      .filter((step) => !step.completed)
      .slice(0, 3)
      .map((step) => step.description),
    conflicts: goal.conflicts,
  }));

  return JSON.stringify(compact, null, 2);
};

const buildPrompt = (payload: EnrichedPayload): string => {
  const intent = payload.intent;
  const context = payload.contextSnippets.join("\n---\n");

  const probability = intent.confidence.toFixed(2);
  const secondary =
    intent.secondBest && intent.secondConfidence
      ? `\nSecondary intent: ${
          intent.secondBest
        } (${intent.secondConfidence?.toFixed(2)})`
      : "";

  const goalSection =
    payload.goalContext && payload.goalContext.length
      ? `\n\n[GOALS]\n${summarizeGoalContext(payload.goalContext)}`
      : "";

  return `"""
[INTENT]
${intent.label} (p=${probability})${secondary}

[SLOTS]
${JSON.stringify(intent.slots, null, 2)}

[CONTEXT]
${context || "n/a"}

[USER]
${payload.userText}

[USER_CONFIG]
${JSON.stringify(payload.userConfig ?? {}, null, 2)}${goalSection}
"""`;
};

type OfflineCapableError = Error & { offline?: boolean };

const isOfflineNetworkError = (error: unknown): boolean => {
  if (!error || typeof error !== "object") {
    return false;
  }
  const message =
    typeof (error as { message?: unknown }).message === "string"
      ? (error as { message: string }).message
      : "";
  return (
    /Failed to fetch/i.test(message) || /Network request failed/i.test(message)
  );
};

const hasOfflineFlag = (error: unknown): error is OfflineCapableError => {
  return Boolean(
    error &&
      typeof error === "object" &&
      "offline" in (error as Record<string, unknown>) &&
      typeof (error as OfflineCapableError).offline === "boolean"
  );
};

const parseToolArguments = (raw: string): PlannerResponse | null => {
  try {
    const parsed = JSON.parse(raw) as PlannerResponse;
    if (!parsed || typeof parsed !== "object") {
      return null;
    }
    if (
      typeof parsed.action !== "string" ||
      ![
        "journal.create",
        "goal.create",
        "schedule.create",
        "reflect",
        "settings.update",
        "noop",
      ].includes(parsed.action)
    ) {
      return null;
    }
    if (parsed.ask !== null && typeof parsed.ask !== "string") {
      return null;
    }
    if (!parsed.payload || typeof parsed.payload !== "object") {
      return null;
    }
    return parsed;
  } catch (error) {
    console.warn("[planner] Failed to parse tool args", error);
    return null;
  }
};

const buildCacheKey = (payload: EnrichedPayload): string => {
  return JSON.stringify({
    label: payload.intent.label,
    slots: payload.intent.slots,
    context: payload.contextSnippets,
    text: payload.userText,
    userConfig: payload.userConfig,
    goalContext: payload.goalContext,
  });
};

const plannerTTL = (action: string | undefined): number => {
  switch (action) {
    case "reflect":
      return 2 * 60 * 1000;
    case "settings.update":
      return 5 * 60 * 1000;
    case "noop":
      return 60 * 1000;
    default:
      return 0;
  }
};

export async function planAction(args: {
  payload: EnrichedPayload;
  apiKey?: string;
}): Promise<PlannerResult> {
  const cacheKey = buildCacheKey(args.payload);
  const cached = await EdgeCache.get<PlannerResponse>(cacheKey);
  if (cached) {
    return {
      response: cached,
      raw: { cached: true },
    };
  }

  const apiKey = resolveOpenAIApiKey(args.apiKey);
  const body = {
    model: resolveModel(),
    temperature: 0,
    tools: [TOOL_DEFINITION],
    tool_choice: {
      type: "function",
      function: { name: "plan_router" },
    },
    messages: [
      { role: "system", content: PLANNER_SYSTEM_PROMPT },
      { role: "user", content: buildPrompt(args.payload) },
    ],
  };

  let data: any;
  try {
    const res = await fetch("https://api.openai.com/v1/chat/completions", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${apiKey}`,
      },
      body: JSON.stringify(body),
    });

    if (!res.ok) {
      const errorBody = await res.text().catch(() => "");
      const error = new Error(
        `Planner request failed: ${res.status} ${errorBody.slice(0, 160)}`
      );
      throw error;
    }

    data = await res.json();
  } catch (error: unknown) {
    if (hasOfflineFlag(error) && error.offline) {
      throw error;
    }

    const offline = isOfflineNetworkError(error);
    const baseMessage =
      error instanceof Error && typeof error.message === "string"
        ? error.message
        : "Planner request failed";
    const offlineError: OfflineCapableError = new Error(
      offline ? "Planner offline" : baseMessage
    );
    offlineError.offline = offline;
    throw offlineError;
  }
  const firstChoice = data?.choices?.[0]?.message;
  const toolCall = firstChoice?.tool_calls?.[0];
  const rawArgs: string | undefined = toolCall?.function?.arguments;
  const parsed = rawArgs ? parseToolArguments(rawArgs) : null;

  if (parsed) {
    const ttl = plannerTTL(parsed.action);
    if (ttl > 0) {
      EdgeCache.set(cacheKey, parsed, ttl).catch((error) => {
        console.warn("[planner] cache set failed", error);
      });
    }
  }

  return {
    response: parsed,
    raw: data,
  };
}

export const PlannerTools = TOOL_DEFINITION;

File: services/python-ai/app/agent/planner.py
import json
from typing import Any, Dict, Optional

from .cache import EdgeCache
from .types import EnrichedPayload, PlannerResponse

PLANNER_SYSTEM_PROMPT = (
    "Planner chooses tools: journal.create|goal.create|schedule.create|reflect|settings.update|noop.\n"
    "Provided state includes structured goals, schedule windows, and constraints.\n"
    "Output JSON is schema-enforced and idempotent. Never hallucinate IDs."
)


def _build_prompt(payload: EnrichedPayload) -> str:
    intent = payload.intent
    context = "\n---\n".join(payload.contextSnippets)
    probability = f"{intent.confidence:.2f}"
    secondary = f"\nSecondary intent: {intent.secondBest} ({intent.secondConfidence:.2f})" if intent.secondBest and intent.secondConfidence else ""
    goal_section = ""
    if payload.goalContext:
        goal_section = "\n\n[GOALS]\n" + json.dumps(
            [
                {
                    "id": goal.id,
                    "title": goal.title,
                    "status": goal.status,
                    "priority": f"{float(goal.priority_score):.2f}",
                    "current_step": goal.current_step,
                    "next_micro_steps": [step.description for step in goal.micro_steps if not step.completed][:3],
                    "conflicts": goal.conflicts,
                }
                for goal in payload.goalContext
            ],
            indent=2,
        )
    return f'"""\n[INTENT]\n{intent.label} (p={probability}){secondary}\n\n[SLOTS]\n{json.dumps(intent.slots, indent=2)}\n\n[CONTEXT]\n{context or "n/a"}\n\n[USER]\n{payload.userText}\n\n[USER_CONFIG]\n{json.dumps(payload.userConfig or {}, indent=2)}{goal_section}\n"""'


def _parse_tool_arguments(raw: str) -> Optional[PlannerResponse]:
    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError:
        return None
    if not isinstance(parsed, dict):
        return None
    if parsed.get("action") not in {"journal.create", "goal.create", "schedule.create", "reflect", "settings.update", "noop"}:
        return None
    if parsed.get("ask") is not None and not isinstance(parsed.get("ask"), str):
        return None
    if not isinstance(parsed.get("payload"), dict):
        return None
    return PlannerResponse(**parsed)


def _build_cache_key(payload: EnrichedPayload) -> str:
    return json.dumps(
        {
            "label": payload.intent.label,
            "slots": payload.intent.slots,
            "context": payload.contextSnippets,
            "text": payload.userText,
            "userConfig": payload.userConfig,
            "goalContext": [goal.model_dump() for goal in payload.goalContext] if payload.goalContext else None,
        },
        sort_keys=True,
    )


def _planner_ttl(action: Optional[str]) -> int:
    if action == "reflect":
        return 2 * 60 * 1000
    if action == "settings.update":
        return 5 * 60 * 1000
    if action == "noop":
        return 60 * 1000
    return 0


def _heuristic_plan(payload: EnrichedPayload) -> PlannerResponse:
    label = payload.intent.label.lower()
    ask: Optional[str] = None
    action = "noop"
    if "schedule" in label:
        action = "schedule.create"
    elif "goal" in label:
        action = "goal.create"
    elif "journal" in label or "entry" in label:
        action = "journal.create"
    elif "reflect" in label:
        action = "reflect"
    elif "command" in label:
        action = "settings.update"
    return PlannerResponse(action=action, ask=ask, payload={"slots": payload.intent.slots, "context": payload.contextSnippets})


async def plan_action(args: Dict[str, Any]) -> Dict[str, Any]:
    payload: EnrichedPayload = args["payload"]
    cache_key = _build_cache_key(payload)
    cached = await EdgeCache.get(cache_key)
    if cached:
        return {"response": PlannerResponse(**cached) if isinstance(cached, dict) else cached, "raw": {"cached": True}}

    prompt = _build_prompt(payload)
    response = _heuristic_plan(payload)

    ttl = _planner_ttl(response.action)
    if ttl > 0:
        await EdgeCache.set(cache_key, response.model_dump(), ttl)

    return {"response": response, "raw": {"prompt": prompt, "heuristic": True}}


PlannerTools = {
    "type": "function",
    "function": {"name": "plan_router", "description": "Select downstream action and structured payload for Riflett subsystems.", "parameters": {}},
}

